### PR TITLE
docs: Add SDK Search reference

### PR DESCRIPTION
**Pull request by tlgimenes**
## What's the purpose of this pull request?
As title says, this PR adds sdk search reference. 

Open this page for testing: `/reference/sdk/search/Overview`

image

### CODE REVIEW

File: packages/sdk/README.md [hunk 1]
```diff
@@ -1,34 +1,3 @@
...
-```
-
-## Analytics
-
-The analytics data layer is based on [the official GA4 specification](https://developers.google.com/gtagjs/reference/ga4-events).
-
-## License
-
-MIT © [VTEX](https://github.com/vtex/faststore)
+Docs moved to https://faststore.dev/reference/sdk/overview
```
Comment (lariciamota) [inline packages/sdk/README.md:3]: What about applying @carolinamenezes' suggestion (in another PR) here also?
```suggestion
For more information, please refer to our documentation https://faststore.dev/reference/sdk/overview
```

File: docs/docs/reference/sdk/search/Overview.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+### useSearchState
+
+The `useSearchState` file defines the interfaces that describe a faceted search state and the functions that change these states.
+
+The main `State` interface allows you to handle the following variables:
+
+|Argument|Type        |Description|
+|--------|------------|---------------------------------------------------|
+|`base`  |`string`    | Base URL path for the search context. This is important when the search context is a subset of what the browser is currently showing to the user, like in an internationalized context where your pages are scoped by a `/locale` param like in `/en` or `/pt`|
+|`page`  |`number`    ||
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/Overview.md:41]: @tlgimenes Can you give a full example for the `base` URL?

Also, we're missing a description for the `page` arg. Is this used for infinite scroll?

File: docs/docs/reference/sdk/search/Overview.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+|`selectedFacets`|`Facet[]`|All selected facets.|
+|`sort`  | `SearchSort`|Selected sorting criteria for the search (e.g, ascending/descending price, number of reviews)|
+|`term`  |`string`    | Terms of the full text search.|
+
+To change the facet state, the following functions are also available:
+- `setFacet` - Select a facet by its key.
+- `toggleFacet` - Replace a selected facet by key.
+- `removeFacet` - Remove the selected facet by its value.
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/Overview.md:49]: @tlgimenes I added the corresponding function names, but I'm not completely if this is correct.

File: docs/docs/reference/sdk/search/Overview.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+The `serializer` file handles URL parsing to create a serialized URL for every facet combination.
+
+### useInfiniteSearchState
+
+The `useInfiniteSearchState` file is an optional function you can use to add infinite scroll in your faceted search pages. The `useInfiniteSearchState` file keeps a record of which pages and products have already been displayed to shoppers with the infinite scroll pagination.
+
+## Example
+
+You can find some examples and boundary conditions [in our tests.](https://github.com/vtex/faststore/tree/master/packages/sdk/test/search)
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/Overview.md:61]: I felt a bit lost with this example. We talked about `useSearchState`, `serializer` and `useInfiniteSearchState`, but I'm not sure about how to use them in this example.

Comment (carolinamenezes) [reply docs/docs/reference/sdk/search/Overview.md:61]: Also, as an Overview doc, I think we need to talk briefly about what is inside this section:
- SearchProvider
- usePagination
- useSearch

File: docs/sidebars.js [hunk 5]
```diff
@@ -93,19 +93,29 @@ module.exports = {
       label: 'SDK',
       collapsed: false,
           items: [
-            'reference/sdk/overview',
+            'reference/sdk/faststore-sdk',
```
Comment (carolinamenezes) [inline docs/sidebars.js:96]:
```suggestion
            'reference/sdk/overview',
```

File: docs/docs/reference/sdk/search/usePagination.mdx [hunk 6]
```diff
@@ -0,0 +1,47 @@
+import PropsComponent from "@site/src/components/PropsComponent/PropsComponent";
+
+# usePagination
+
+This hooks gives you the links to next/prev pages on the search pagination.
+This is helpfull for creating pagination markers like `<link rel="next"/>` and `<link rel="prev"/>` tags and buttons in the UI
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/usePagination.mdx:6]:
```suggestion
The `usePagination` hook provides the pagination links (i.e., `next`/`prev`) of a search result. You can use the `rel="next"` and `rel="prev"` links along with the `<link />` tag to add navigation buttons to your search result pages.
```

File: docs/docs/reference/sdk/search/usePagination.mdx [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      {next && <a href={next.link}>Next Page</a>}
+    </>
+  )
+}
+```
+
+## Props
+<PropsComponent
+  type="number"
+  description="Total number of products in the search result"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/usePagination.mdx:33]:
```suggestion
  description="Total number of products returned by the search query."
```

File: docs/docs/reference/sdk/search/usePagination.mdx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+/>
+
+## Return
+<PropsComponent
+    name="prev"
+    type="{ cursor: number; link: string} | false"
+    description="Link and cursor for previous page on search pagination"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/usePagination.mdx:40]:
```suggestion
    description="Cursor position and corresponding link for the previous page of the search result."
```

File: docs/docs/reference/sdk/search/usePagination.mdx [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+/>
+
+<PropsComponent
+    name="next"
+    type="{ cursor: number; link: string} | false"
+    description="Link and cursor for next page on search pagination"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/usePagination.mdx:46]:
```suggestion
    description="Cursor position and corresponding link for the next page of the search result."
```

File: docs/docs/reference/sdk/search/useSearch.mdx [hunk 10]
```diff
@@ -0,0 +1,115 @@
+import PropsComponent from "@site/src/components/PropsComponent/PropsComponent";
+
+# useSearch
+
+This hook allows you to consume `SearchContext`. This context allows you to apply changes to the search state as well as consume its values
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/useSearch.mdx:5]:
```suggestion
The `useSearch` hook consumes the search context provided by the `SearchProvider` component. You can use this context to change the search state and consume its values.
```

File: docs/docs/reference/sdk/search/useSearch.mdx [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+  )
+}
+```
+
+## Return
+
+<PropsComponent
+    name="itemsPerPage"
+    type="number"
+    description="Number of items in a search page"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/useSearch.mdx:57]:
```suggestion
    description="Number of items per page."
```

File: docs/docs/reference/sdk/search/useSearch.mdx [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+/>
+
+<PropsComponent
+    name="state"
+    type="SearchState"
+    description="Current state of search"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/useSearch.mdx:63]:
```suggestion
    description="Current state of the facets."
```

File: docs/docs/reference/sdk/search/useSearch.mdx [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+/>
+
+<PropsComponent
+    name="setSort"
+    type="(sort: SearchSort) => void"
+    description="Set the sorting option of the search"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/useSearch.mdx:69]:
```suggestion
    description="Sets the sorting criteria of the search results."
```

File: docs/docs/reference/sdk/search/useSearch.mdx [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+/>
+
+<PropsComponent
+    name="setTerm"
+    type="(term: string | null) => void"
+    description="Set the full text term"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/useSearch.mdx:75]:
```suggestion
    description="Sets the full-text term."
```

File: docs/docs/reference/sdk/search/useSearch.mdx [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+/>
+
+<PropsComponent
+    name="setPage"
+    type="(page: number) => void"
+    description="Set the current page on search pagination"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/useSearch.mdx:81]:
```suggestion
    description="Sets the current page on search pagination."
```

File: docs/docs/reference/sdk/search/useSearch.mdx [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+/>
+
+<PropsComponent
+    name="setFacet"
+    type="(facet: Facet, unique?: boolean) => void"
+    description="Select a facet"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/useSearch.mdx:87]:
```suggestion
    description="Selects a specific facet."
```

File: docs/docs/reference/sdk/search/useSearch.mdx [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+/>
+
+<PropsComponent
+  name="removeFacet"
+  type="(facet: Facet) => void"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/useSearch.mdx:92]:
```suggestion
  type="(facet: Facet) => void"
  description="Removes a specific facet."
```

File: docs/docs/reference/sdk/search/useSearch.mdx [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+/>
+
+<PropsComponent
+  name="toggleFacet"
+  type="(facet: Facet) => void"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/useSearch.mdx:97]:
```suggestion
  type="(facet: Facet) => void"
  description="Replaces a specific facet."
```

Comment (carolinamenezes) [reply docs/docs/reference/sdk/search/useSearch.mdx:97]: Não entendi esse.

File: docs/docs/reference/sdk/search/useSearch.mdx [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+/>
+
+<PropsComponent
+  name="toggleFacets"
+  type="(facets: Facet[]) => void"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/useSearch.mdx:102]: ?

File: docs/docs/reference/sdk/search/useSearch.mdx [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+/>
+
+<PropsComponent
+  name="addPrevPage"
+  type="() => void"
+  description="Prepend a page to infinite scroll pagination"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/useSearch.mdx:108]:
```suggestion
  description="Prepends a page to the infinite scroll pagination."
```

File: docs/docs/reference/sdk/search/useSearch.mdx [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+/>
+
+<PropsComponent
+  name="addNextPage"
+  type="() => void"
+  description="Append a page to infinite scroll pagination"
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/useSearch.mdx:114]:
```suggestion
  description="Appends a page to the infinite scroll pagination."
```

File: docs/docs/reference/sdk/search/SearchProvider.mdx [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+<PropsComponent
+    name="base"
+    type="string"
+    description="Base path of the search page. Usefull when dealing with localization and prefixing paths by the locale, e.g.: /pt-br/"
+/>
+
+<PropsComponent
+    name="selectedFacets"
+    type="Array<{ key: string, value: string }>"
+    description="Array of selected facets on the search. Make sure NOT to include personalization facets, like channel or price tables"
```
[the remark this answers is not in the document]
Comment (tlgimenes) [reply docs/docs/reference/sdk/search/SearchProvider.mdx:70]: Sao o sales channel ou price table. Se incluir eles ai, eles vao acabar na URL. Se voce mandar essa url para outra pessoa, pode dar um erro pois essa outra pessoa nao necessariamente vai estar no mesmo sales channel de voce e vai ver outra coisa.

File: docs/docs/reference/sdk/search/useSearch.mdx (hunk 12 above)
Comment (tlgimenes) [reply docs/docs/reference/sdk/search/useSearch.mdx:63]: Not only the facets, but also sort, basePath etc

File: docs/docs/reference/sdk/search/Overview.md (hunk 3 above)
Comment (tlgimenes) [reply docs/docs/reference/sdk/search/Overview.md:49]: it's ok

File: docs/docs/reference/sdk/search/SearchProvider.mdx [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+  itemsPerPage={2}
+  sort="score_desc"
+  term={null}
+  page={0}
+  base=""
+  selectedFacets={}
+>
+  {children}
+</SearchProvider>
+```
```
[the remark this answers is not in the document]
Comment (tlgimenes) [reply docs/docs/reference/sdk/search/SearchProvider.mdx]: what about now?

File: docs/docs/reference/sdk/search/Overview.md (hunk 2 above)
Comment (tlgimenes) [reply docs/docs/reference/sdk/search/Overview.md]: what about now?

File: docs/docs/reference/sdk/search/Overview.md (hunk 4 above)
Comment (tlgimenes) [reply docs/docs/reference/sdk/search/Overview.md]: Ok. I did some refactoring on this part, tell me what you think about it

File: docs/docs/reference/sdk/search/SearchProvider.mdx (hunk 22 above)
[the remark this answers is not in the document]
Comment (igorbrasileiro) [reply docs/docs/reference/sdk/search/SearchProvider.mdx:74]: Mas acho que isso é normal. O site da amazon é assim, por exemplo: buscar por X
- uma conta com prime e frete gratis do prime, o produto X1 fica com menor preço, então ele é o escolhido;
- uma conta sem prime e produto com frete, o produto X2 fica com menor preço, então ele é o escolhido;

Se o usuário 1 compartilhar o link pra o usuário 2, o usuário dois, vai ver X2 e não X1.

File: docs/docs/reference/sdk/search/usePagination.mdx [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+
+```tsx
+function Component () {
+  const totalProducts = 100
+  const { next, prev } = usePagination(totalProducts)
+
+  return (
+    <>
+      {prev && <a href={prev.link}>Previous Page</a>}
+      {next && <a href={next.link}>Next Page</a>}
```
Comment (igorbrasileiro) [inline docs/docs/reference/sdk/search/usePagination.mdx:23]: Recommend adding key props to the example.

File: docs/docs/reference/ui/molecules/Alert.mdx [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+      backgroundColor: '#fff2d4',
+      color: "#eea236",
+      borderRadius: "0.5rem",
+      padding: "1rem"
+    },
+  }
+
+  return (
+    <Alert style={styles.warning}>
+    <React.Fragment key=".0">
```
Comment (igorbrasileiro) [inline docs/docs/reference/ui/molecules/Alert.mdx:36]: To be more consistent, i suggest to use the fragment shorthand `<> / </>`

**Review by igorbrasileiro — APPROVED** [review]
Great Job, @tlgimenes! 
Take care, that some changes from other PRs are being repeated here.

File: docs/docs/reference/sdk/search/SearchProvider.mdx (hunk 22 above)
[the remark this answers is not in the document]
Comment (tlgimenes) [reply docs/docs/reference/sdk/search/SearchProvider.mdx:74]: sim, mas isso vai dar todo o tipo de bug na nossa implementacao. Vai ter duas fontes de verdade de sales channel etc e cada componente vai mostrar uma coisa pois cada um vai consumir esse dado de um provedor de contexto diferente, entao é melhor nao fazermos essa feature pelo menos por enquanto

File: docs/docs/reference/sdk/search/usePagination.mdx (hunk 24 above)
Comment (tlgimenes) [reply docs/docs/reference/sdk/search/usePagination.mdx:23]: key? for what?

File: docs/docs/reference/sdk/search/useSearch.mdx (hunk 12 above)
Comment (carolinamenezes) [reply docs/docs/reference/sdk/search/useSearch.mdx:63]: Is it correct now?

**Review by carolinamenezes — APPROVED** [review]

File: docs/docs/reference/sdk/search/Overview.md [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+  )
+}
+```
+
+The idea here is to use the browser's URL to keep the state of the faceted search. Whenever a new facet is toggled, we navigate the user to this new URL, where the components will re-render, updating the UI. 
+
+This example has two components. One that is the page itself and a `Checkbox`. The page component uses `parseSearchState` for parsing the search state from the url and using the output into `SearchProvider`. 
+Also, `Page` configures `SearchProvider` for displaying `12` products per page and creates an inline `onChange` function. This function is called whenever someone clicks on the `Checkbox`, and navigates the user to the new url with the facets applied. 
+
+For a full implementation, please refer to our official starters. Also, you can find more examples and boundary conditions [in our tests.](https://github.com/vtex/faststore/tree/master/packages/sdk/test/search)
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/Overview.md:90]:
```suggestion
This example has two components: `Page` and `Checkbox`. 

The `Page` component parses the search state from the browser's URL using `parseSearchState` and provides the parsed output to the `SearchProvider` component. The `SearchProvider` wraps the entire page, providing the necessary context for the Faceted Search.

In this example, the inline `onChange` function defined for the `SearchProvider` component is called whenever someone selects the `Checkbox` to apply a new facet. This function uses the parsed output from `parseSearchState` to redirect users to the new URL corresponding to the recently applied facets. Also, the `itemsPerPage` prop is configured to make the `SearchProvider` display 12 products per page.

For practical examples, please refer to our official starters. Also, you can find more examples and boundary conditions [in our tests.](https://github.com/vtex/faststore/tree/master/packages/sdk/test/search)
```

File: docs/docs/reference/sdk/search/Overview.md [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+- `setFacet` - Select a facet by its key.
+- `toggleFacet` - Replace a selected facet by key.
+- `removeFacet` - Remove the selected facet by its value.
+
+### serializer
+
+The `serializer` file handles URL parsing to create a serialized URL for every facet combination.
+
+## Example
+Suppose we want to create a full text search page that renders a checkbox. This checkbox toggles a facet for the brand `faststore`.
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/Overview.md:55]:
```suggestion
Suppose we want to create a full-text search page that renders a checkbox, which toggles a facet for the `faststore` brand. To build this feature, we will use the browser's URL to keep the state of the faceted search. Then, whenever someone selects the `faststore` facet, we will redirect them to the corresponding URL, updating the UI and re-rendering the necessary components to display all product items from the `faststore` brand.
```

File: docs/docs/reference/sdk/search/SearchProvider.mdx [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+/>
+
+<PropsComponent
+    name="selectedFacets"
+    type="Array<{ key: string, value: string }>"
+    description="Array of selected facets on the search."
+/>
+
+:::caution WIP
+Make sure NOT to include personalization facets, like channel or price tables on selectedFacets. If you do, these facets, alongside their values, will be serialized in the url. This means that a user may end up having two channels, one from the Session context and another from the url, causing all sorts of weird bugs
```
Comment (carolinamenezes) [inline docs/docs/reference/sdk/search/SearchProvider.mdx:78]:
```suggestion
:::caution
Don't include personalization facets, such as sales channel and price tables, on the `selectedFacets` property. If you do so, users may end up having two channels: one from the Session context and another from the URL, which may cause all sorts of unexpected bugs.
```

**Review by carolinamenezes — APPROVED** [review]
